base/compiler/typeinfer.jl

```diff
@@ -704,52 +664,54 @@ function type_annotate!(interp::AbstractInterpreter, sv::InferenceState, run_opt
-    # 5. eliminate GotoIfNot if either branch target is unreachable
+    # 4. eliminate GotoIfNot if either or both branches are statically unreachable
     changemap = nothing # initialized if there is any dead region
     for i = 1:nstmt
         expr = stmts[i]
         if was_reached(sv, i)
             if run_optimizer
-                if isa(expr, GotoIfNot) && widenconst(argextype(expr.cond, src, sv.sptypes)) === Bool
+                if isa(expr, GotoIfNot)
                     # 5: replace this live GotoIfNot with:
-                    # - GotoNode if the fallthrough target is unreachable
-                    # - no-op if the branch target is unreachable
-                    if !was_reached(sv, i+1)
-                        expr = GotoNode(expr.dest)
-                    elseif !was_reached(sv, expr.dest)
-                        expr = nothing
+                    # - no-op if :nothrow and the branch target is unreachable
+                    # - cond if :nothrow and both targets are unreachable
+                    # - typeassert if must-throw
+                    if widenconst(argextype(expr.cond, src, sv.sptypes)) === Bool
+                        block = block_for_inst(sv.cfg, i)
+                        if !was_reached(sv, i+1)
+                            cfg_delete_edge!(sv.cfg, block, block + 1)
+                            expr = GotoNode(expr.dest)
+                        elseif !was_reached(sv, expr.dest)
+                            cfg_delete_edge!(sv.cfg, block, block_for_inst(sv.cfg, expr.dest))
+                            expr = nothing
+                        end
+                    elseif ssavaluetypes[i] === Bottom
+                        block = block_for_inst(sv.cfg, i)
+                        cfg_delete_edge!(sv.cfg, block, block + 1)
+                        cfg_delete_edge!(sv.cfg, block, block_for_inst(sv.cfg, expr.dest))
+                        expr = Expr(:call, Core.typeassert, expr.cond, Bool)
                     end
                 end
             end
-            stmts[i] = annotate_slot_load!(interp, undefs, i, sv, expr) # 1&2
-            ssavaluetypes[i] = widenslotwrapper(ssavaluetypes[i]) # 4
+            stmts[i] = expr
+            ssavaluetypes[i] = widenslotwrapper(ssavaluetypes[i]) # 3
         else # i.e. any runtime execution will never reach this statement
             any_unreachable = true
             if is_meta_expr(expr) # keep any lexically scoped expressions
-                ssavaluetypes[i] = Any # 4
+                ssavaluetypes[i] = Any # 3
             else
-                ssavaluetypes[i] = Bottom # 4
+                ssavaluetypes[i] = Bottom # 3
                 stmts[i] = Const(expr) # annotate that this statement actually is dead
             end
         end
     end
 
-    # finish marking used-undef variables
-    for j = 1:nslots
-        if undefs[j]
-            slotflags[j] |= SLOT_USEDUNDEF | SLOT_STATICUNDEF
```

Comment:
The optimizer no longer relies on these flags, but will external consumers need them?

Reply:
We'll find out. I'm somewhat sceptical, but if they are, let's wean them off it.